File: client/src/services/api/aiChats.ts
import { fetchJson, authHeaders, withBase } from './http';

export interface AIChatSummary {
  id: string;
  title: string;
  model: string;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface AIChatCostBreakdown {
  input: string;
  cached: string;
  output: string;
  tool: string;
}

export interface AIChatMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  created_at?: string | null;
  tokens?: number | null;
  cost?: string | null;
  input_tokens?: number | null;
  cached_tokens?: number | null;
  tool_tokens?: number | null;
  total_tokens?: number | null;
  cost_breakdown?: AIChatCostBreakdown | null;
  cache_used?: boolean;
}

export interface AIChatEstimate {
  estimated_input_tokens: number;
  estimated_output_tokens: number;
  estimated_cached_tokens: number | null;
  estimated_cost: string;
  estimated_max_cost: string;
  cost_breakdown?: AIChatCostBreakdown | null;
  currency: string;
  is_estimate: boolean;
}

export interface AIUsageSummary {
  period: string;
  input_tokens: number;
  cached_tokens: number;
  output_tokens: number;
  tool_tokens: number;
  total_tokens: number;
  cost: string;
  currency: string;
  request_count: number;
}

export interface AIChatStreamOptions {
  content: string;
  signal?: AbortSignal;
  onChunk: (text: string) => void;
  onUsage?: (usage: AIChatStreamUsage) => void;
  onDone: () => void;
  onError: (error: Error) => void;
}

export interface AIChatStreamUsage {
  message_uid: string;
  prompt_tokens: number;
  cached_tokens: number;
  completion_tokens: number;
  tool_tokens: number;
  total_tokens: number;
  cost: string;
  cost_breakdown?: AIChatCostBreakdown | null;
  currency: string;
  deducted: boolean;
  deduct_error: string | null;
}

export async function listAIChats(limit = 50): Promise<AIChatSummary[]> {
  const data = await fetchJson<{ chats: AIChatSummary[] }>(`/api/ai/chats?limit=${limit}`);
  return data.chats ?? [];
}

export async function createAIChat(title = ''): Promise<AIChatSummary> {
  return fetchJson<AIChatSummary>('/api/ai/chats', {
    method: 'POST',
    body: JSON.stringify({ title }),
  });
}

export async function getAIChatMessages(chatId: string, limit = 100): Promise<AIChatMessage[]> {
  const data = await fetchJson<{ messages: AIChatMessage[] }>(
    `/api/ai/chats/${encodeURIComponent(chatId)}/messages?limit=${limit}`,
  );
  return data.messages ?? [];
}

export async function estimateAIChatMessage(
  chatId: string,
  messages: { role: string; content: string }[],
  maxOutputTokens?: number,
): Promise<AIChatEstimate> {
  return fetchJson<AIChatEstimate>(`/api/ai/chats/${encodeURIComponent(chatId)}/messages/estimate`, {
    method: 'POST',
    body: JSON.stringify({ messages, max_output_tokens: maxOutputTokens }),
  });
}

export async function getAIUsageSummary(period: 'current' | 'previous' = 'current'): Promise<AIUsageSummary> {
  return fetchJson<AIUsageSummary>(`/api/ai/usage/summary?period=${period}`);
}

/**
 * Потоковая отправка сообщения в серверный AI-чат (SSE через `/api/ai/chats/{id}/messages`).
 */
export async function streamAIChatMessage(chatId: string, opts: AIChatStreamOptions): Promise<void> {
  const { content, signal, onChunk, onUsage, onDone, onError } = opts;
  try {
    const response = await fetch(withBase(`/api/ai/chats/${encodeURIComponent(chatId)}/messages`), {
      method: 'POST',
      headers: authHeaders({ 'Content-Type': 'application/json' }),
      body: JSON.stringify({ content }),
      signal,
    });
    if (!response.ok || !response.body) {
      const text = await response.text().catch(() => '');
      throw new Error(
        `HTTP ${response.status} ${response.statusText}${text ? `: ${text.slice(0, 300)}` : ''}`,
      );
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';

    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      const lines = buffer.split('\n');
      buffer = lines.pop() ?? '';
      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed.startsWith('data:')) continue;
        const data = trimmed.slice(5).trim();
        if (!data || data === '[DONE]') continue;
        try {
          const event = JSON.parse(data);
          if (event.type === 'chunk' && typeof event.content === 'string') {
            onChunk(event.content);
          } else if (event.type === 'usage' && onUsage) {
            onUsage(event as AIChatStreamUsage);
          } else if (event.type === 'error') {
            throw new Error(event.message || 'AI gateway error');
          }
        } catch (err) {
          if (err instanceof Error && !(err instanceof DOMException)) {
            onError(err);
            return;
          }
        }
      }
    }

    onDone();
  } catch (error) {
    if (error instanceof DOMException && error.name === 'AbortError') {
      onDone();
      return;
    }
    onError(error instanceof Error ? error : new Error(String(error)));
  }
}

File: api/src/routers/ai_chats.py
"""
Layer: Interface Adapters — Web (Routers)
Package: src.routers.ai_chats
Responsibility: HTTP-контроллеры персистентных AI-чатов (учёт токенов/стоимости).
Все мутации требуют авторизации; ownership проверяется в use cases.

Маршруты:
  GET    /api/ai/chats                — список чатов пользователя
  POST   /api/ai/chats                — создать чат
  GET    /api/ai/chats/{id}/messages  — история сообщений
  POST   /api/ai/chats/{id}/messages  — отправить сообщение (SSE-стрим + usage)
  POST   /api/ai/chats/{id}/messages/estimate — оценка токенов/стоимости
  GET    /api/ai/usage/summary        — агрегат usage за период
"""
import json
import logging
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.responses import JSONResponse, StreamingResponse
from pydantic import BaseModel, Field

from application.ai_chats.create_chat import create_ai_chat
from application.ai_chats.estimate_message import estimate_message
from application.ai_chats.get_chat import get_ai_chat, get_ai_chat_messages
from application.ai_chats.list_chats import list_ai_chats
from application.ai_chats.send_message import send_ai_message_stream
from application.ai_chats.usage_summary import usage_summary
from domain.exceptions import AuthorizationFailed, NotFoundError
from domain.rules.ai_pricing import calculate_usage_cost
from web.dependencies import (
    get_ai_chat_repository,
    get_ai_gateway,
    get_billing_client,
    get_current_user,
    get_deepseek_tokenizer,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/ai", tags=["ai_chats"])


class CreateChatRequest(BaseModel):
    title: str = Field(default="", max_length=200)


class EstimateRequest(BaseModel):
    messages: List[dict] = Field(default_factory=list)
    max_output_tokens: Optional[int] = Field(default=None, ge=1, le=1_048_576)


class SendMessageRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=100_000)


@router.get("/chats")
async def list_chats(
    limit: int = 50,
    user: dict = Depends(get_current_user),
    repository=Depends(get_ai_chat_repository),
):
    chats = list_ai_chats(repository=repository, user_uid=user.get("uid", ""), limit=limit)
    return {
        "chats": [
            {
                "id": chat.uid,
                "title": chat.title,
                "model": chat.model,
                "created_at": chat.created_at.isoformat() if chat.created_at else None,
                "updated_at": chat.updated_at.isoformat() if chat.updated_at else None,
            }
            for chat in chats
        ]
    }


@router.post("/chats")
async def create_chat(
    body: CreateChatRequest,
    user: dict = Depends(get_current_user),
    repository=Depends(get_ai_chat_repository),
):
    chat = create_ai_chat(
        repository=repository,
        user_uid=user.get("uid", ""),
        title=body.title,
    )
    return {
        "id": chat.uid,
        "title": chat.title,
        "model": chat.model,
        "created_at": chat.created_at.isoformat() if chat.created_at else None,
    }


@router.get("/chats/{chat_id}/messages")
async def get_messages(
    chat_id: str,
    limit: int = 100,
    user: dict = Depends(get_current_user),
    repository=Depends(get_ai_chat_repository),
):
    try:
        messages = get_ai_chat_messages(
            repository=repository,
            chat_uid=chat_id,
            user_uid=user.get("uid", ""),
            limit=limit,
        )
    except NotFoundError:
        raise HTTPException(status_code=404, detail="chat_not_found")
    except AuthorizationFailed:
        raise HTTPException(status_code=403, detail="forbidden")

    usages = repository.list_usage_for_chat(chat_id, limit=limit)
    usage_by_message = {u.message_uid: u for u in usages}
    # usage привязан к сообщению ассистента; входная часть (input+кэш) проецируется
    # на парное сообщение пользователя, выходная (output+инструменты) — на ассистента.
    return {"messages": [_message_payload(m, usage_by_message, messages) for m in messages]}


def _message_payload(m, usage_by_message: dict, messages) -> dict:
    """Payload сообщения с учётом фактического usage (input-часть на user, output — на assistant)."""
    base = {
        "id": m.uid,
        "role": m.role,
        "content": m.content,
        "created_at": m.created_at.isoformat() if m.created_at else None,
        "tokens": None,
        "cost": None,
        "input_tokens": None,
        "cached_tokens": None,
        "tool_tokens": None,
        "total_tokens": None,
        "cost_breakdown": None,
        "cache_used": False,
    }
    usage = usage_by_message.get(m.uid)
    if m.role == "user":
        # парное сообщение ассистента — следующее в истории
        idx = next((i for i, x in enumerate(messages) if x.uid == m.uid), -1)
        nxt = messages[idx + 1] if idx >= 0 and idx + 1 < len(messages) else None
        if nxt is not None and nxt.role == "assistant":
            usage = usage_by_message.get(nxt.uid)
    if usage is None:
        return base

    if m.role == "user":
        cost = calculate_usage_cost(
            input_tokens=usage.actual_input_tokens,
            cached_input_tokens=usage.actual_cached_tokens,
            output_tokens=0,
            tool_tokens=0,
        )
        base.update(
            {
                "tokens": usage.actual_input_tokens,
                "cost": str(cost.total.normalize()),
                "input_tokens": usage.actual_input_tokens,
                "cached_tokens": usage.actual_cached_tokens,
                "tool_tokens": 0,
                "total_tokens": usage.actual_input_tokens,
                "cost_breakdown": {
                    "input": str(cost.input_cost.normalize()),
                    "cached": str(cost.cached_input_cost.normalize()),
                    "output": "0",
                    "tool": "0",
                },
                "cache_used": bool(usage.actual_cached_tokens),
            }
        )
    else:
        cost = calculate_usage_cost(
            input_tokens=0,
            cached_input_tokens=0,
            output_tokens=usage.actual_output_tokens,
            tool_tokens=usage.actual_tool_tokens,
        )
        base.update(
            {
                "tokens": usage.actual_output_tokens,
                "cost": str(cost.total.normalize()),
                "input_tokens": 0,
                "cached_tokens": 0,
                "tool_tokens": usage.actual_tool_tokens,
                "total_tokens": usage.actual_output_tokens + usage.actual_tool_tokens,
                "cost_breakdown": {
                    "input": "0",
                    "cached": "0",
                    "output": str(cost.output_cost.normalize()),
                    "tool": str(cost.tool_cost.normalize()),
                },
                "cache_used": False,
            }
        )
    return base


@router.post("/chats/{chat_id}/messages/estimate")
async def estimate(
    chat_id: str,
    body: EstimateRequest,
    user: dict = Depends(get_current_user),
    repository=Depends(get_ai_chat_repository),
    tokenizer=Depends(get_deepseek_tokenizer),
):
    try:
        get_ai_chat(repository=repository, chat_uid=chat_id, user_uid=user.get("uid", ""))
    except NotFoundError:
        raise HTTPException(status_code=404, detail="chat_not_found")
    except AuthorizationFailed:
        raise HTTPException(status_code=403, detail="forbidden")
    result = estimate_message(
        tokenizer=tokenizer,
        messages=body.messages,
        max_output_tokens=body.max_output_tokens,
    )
    return result


@router.post("/chats/{chat_id}/messages")
async def send_message(
    chat_id: str,
    body: SendMessageRequest,
    user: dict = Depends(get_current_user),
    repository=Depends(get_ai_chat_repository),
    tokenizer=Depends(get_deepseek_tokenizer),
    gateway=Depends(get_ai_gateway),
    billing=Depends(get_billing_client),
):
    """Отправляет сообщение в AI-чат и стримит ответ (SSE)."""

    async def event_stream():
        async for event in send_ai_message_stream(
            repository=repository,
            tokenizer=tokenizer,
            gateway=gateway,
            billing=billing,
            chat_uid=chat_id,
            user_uid=user.get("uid", ""),
            content=body.content,
        ):
            yield f"data: {json.dumps(event, ensure_ascii=False)}\n\n"

    # Проверка существования/владельца до начала стрима.
    try:
        get_ai_chat(repository=repository, chat_uid=chat_id, user_uid=user.get("uid", ""))
    except NotFoundError:
        raise HTTPException(status_code=404, detail="chat_not_found")
    except AuthorizationFailed:
        raise HTTPException(status_code=403, detail="forbidden")

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )


@router.get("/usage/summary")
async def usage_summary_route(
    period: str = "current",
    user: dict = Depends(get_current_user),
    repository=Depends(get_ai_chat_repository),
):
    if period not in ("current", "previous"):
        raise HTTPException(status_code=400, detail="invalid_period")
    return usage_summary(
        repository=repository,
        user_uid=user.get("uid", ""),
        period=period,
    )
